### PR TITLE
[dynamo] don't use LazyModuleMixin.cls_to_become if it is None

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #99943

**TL;DR**: This PR fixes handling for lazy modules where `cls_to_become is None`. In those cases, we should leave the type of the lazy module as the old value.

**Details**:
Lazy modules are intended to be initialized at execution; some of them are also supposed to switch to a different type after they have been initialized. However, not all are supposed to switch; see this logic from `nn/modules/lazy.py`

```python
    def _infer_parameters(self, ...):
        ...
        if module.cls_to_become is not None:
            module.__class__ = module.cls_to_become
```

i.e., we should leave the module type as the old value if `module.cls_to_become is None`. This PR updates dynamo's handling to match this behavior.

Test `test_lazy_module_no_cls_to_become` added to `test/dynamo/test_module.py`.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

Differential Revision: [D45253698](https://our.internmc.facebook.com/intern/diff/D45253698)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45253698).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45253698).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45253698).

**Review by jansel — APPROVED** [review]

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45253698).

**Comment by davidberard98** [comment]
somehow, adding this test causes `test_backward_hooks`, `test_hooks_allowed_modules` to fail with `RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn`. Investigating...

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45253698).

**Comment by davidberard98** [comment]
Isolated that issue in https://github.com/pytorch/pytorch/issues/100001 with help from @wconstab . It appears unrelated to this PR (I can repro on the main branch without this PR).
